bep14.go

```
@@ -0,0 +1,476 @@
+package torrent
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+
+	"github.com/anacrolix/missinggo"
+	"github.com/anacrolix/torrent/metainfo"
+)
+
+// http://bittorrent.org/beps/bep_0014.html
+// TODO http://bittorrent.org/beps/bep_0026.html
+
+const (
+	bep14_host4    = "239.192.152.143:6771"
+	bep14_host6    = "[ff15::efc0:988f]:6771"
+	bep14_announce = "BT-SEARCH * HTTP/1.1\r\n" +
+		"Host: %s\r\n" +
+		"Port: %s\r\n" +
+		"%s" +
+		"\r\n" +
+		"\r\n"
+	bep14_announce_infohash = "Infohash: %s\r\n"
+	bep14_long_timeout      = 1 * time.Minute
+	bep14_short_timeout     = 2 * time.Second // bep14 - 1 minute. not practial. what if use start/stop another torrent? so make it 2 secs.
+	bep14_max               = 0               // maximum hashes per request, 0 - only limited by udp packet size
+)
+
+type LPDConn struct {
+	stop  missinggo.Event
+	force missinggo.Event
+
+	lpd         *LPDServer
+	network     string // "udp4" or "udp6"
+	addr        *net.UDPAddr
+	mcListener  *net.UDPConn
+	mcPublisher *net.UDPConn
+	host        string // bep14_host4 or bep14_host6
+	closed      bool
+}
+
+func setMulticastInterface(m *LPDConn, iface *net.Interface) error {
+	if m.network == "udp4" {
+		p := ipv4.NewPacketConn(m.mcPublisher)
+		if err := p.SetMulticastInterface(iface); err != nil {
+			log.Printf("Set multicast interface error: %v\n", err)
+			return err
+		}
+	}
+	if m.network == "upd6" {
+		p := ipv6.NewPacketConn(m.mcPublisher)
+		if err := p.SetMulticastInterface(iface); err != nil {
+			log.Printf("Set multicast interface error: %v\n", err)
+			return err
+		}
+	}
+	return nil
```

Maybe return an error for unexpected network?

---

hmm, i haven't found an elegant way to do this, without using `m.network != "udp4" && m.network != "udp6"`

now that you draw my attention to it, i'm curious why the p variable is not garbage collected... i wanted to add it to the lpdConn struct but it has type `*ipv4.PacketConn` or `*ipv6.PacketConn`

what do you think?

---

You shouldn't rely on GC to clean up resources in Go. In most code depending on stdlib, os.File has a special handler to clean up because people forget so often. I expect the ipv4/6 PacketConn aren't doing this underneath or something.

Try to put those into a common interface type? Or you can make an interface { a | b } type to hold it possibly.